{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gap Annotation to Bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>a</td>\n",
       "      <td>.</td>\n",
       "      <td>BAC</td>\n",
       "      <td>11</td>\n",
       "      <td>12</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gap00001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>a</td>\n",
       "      <td>.</td>\n",
       "      <td>BAC</td>\n",
       "      <td>19</td>\n",
       "      <td>22</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gap00002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>a</td>\n",
       "      <td>.</td>\n",
       "      <td>BAC</td>\n",
       "      <td>32</td>\n",
       "      <td>35</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gap00003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>b</td>\n",
       "      <td>.</td>\n",
       "      <td>BAC</td>\n",
       "      <td>8</td>\n",
       "      <td>10</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gap00001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>b</td>\n",
       "      <td>.</td>\n",
       "      <td>BAC</td>\n",
       "      <td>11</td>\n",
       "      <td>12</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gap00002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>b</td>\n",
       "      <td>.</td>\n",
       "      <td>BAC</td>\n",
       "      <td>21</td>\n",
       "      <td>25</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gap00003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>b</td>\n",
       "      <td>.</td>\n",
       "      <td>BAC</td>\n",
       "      <td>34</td>\n",
       "      <td>38</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gap00004</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  name source type  start  end score strand phase    attribute\n",
       "0    a      .  BAC     11   12     .      +     .  ID=gap00001\n",
       "1    a      .  BAC     19   22     .      +     .  ID=gap00002\n",
       "2    a      .  BAC     32   35     .      +     .  ID=gap00003\n",
       "3    b      .  BAC      8   10     .      +     .  ID=gap00001\n",
       "4    b      .  BAC     11   12     .      +     .  ID=gap00002\n",
       "5    b      .  BAC     21   25     .      +     .  ID=gap00003\n",
       "6    b      .  BAC     34   38     .      +     .  ID=gap00004"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import HTSeq\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "def single_sequence(sequence):\n",
    "    '''\n",
    "    sequence       refers to a HTSeq.Sequence object\n",
    "    \n",
    "        the function returns DataFrame object in GFF3 format \n",
    "    '''\n",
    "    seq = sequence.seq.decode()\n",
    "    \n",
    "    bases = ['A','a','C','c','G','g','T','t']\n",
    "\n",
    "    start = []\n",
    "    end = []\n",
    "\n",
    "    above = None\n",
    "    for i in range(len(seq)):\n",
    "        if seq[i] in bases and above == 'N':\n",
    "            end.append(i+1)\n",
    "        elif seq[i] == 'N' and above != 'N':\n",
    "            start.append(i+1)\n",
    "        above = seq[i]\n",
    "\n",
    "    if seq[-1] == 'N':\n",
    "        end.append(len(seq))\n",
    "        \n",
    "    gff3 = pd.DataFrame({'name':sequence.name, 'source':'.','type':'BAC', \\\n",
    "                         'start':start, 'end':end, 'score':'.', \\\n",
    "                         'strand':'+', 'phase':'.', 'attribute':['ID=gap%05d' % i for i in range(1,len(start)+1)]})\n",
    "    \n",
    "    return gff3\n",
    "\n",
    "def annotateGap(genome_file, gff3_file):\n",
    "    GFF3 = pd.DataFrame()\n",
    "    for i in HTSeq.FastaReader(genome_file):\n",
    "        gff3 = single_sequence(i)\n",
    "        GFF3 = pd.concat([GFF3,gff3],ignore_index=True)\n",
    "        \n",
    "    GFF3.to_csv(gff3_file, sep='\\t', header=None, index=None)\n",
    "    return GFF3\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    annotateGap('seq.fa','seq.gff3')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
