# Genomic Resources of Broomcorn Millet: Demonstration and Application of a High-throughput BAC Mapping Pipeline
## Background
With high-efficient water-use and drought tolerance, broomcorn millet has emerged as candidate for food security. To promote its research process for molecular breeding and functional research, a comprehensive genome resources is of great important. 
## Results
Herein, we constructed the first BAC library for broomcorn millet, generated BAC end sequences and integrated BAC clones into genome by a novel pipeline for BAC end profiling depending on clone-array pooled shotgun sequencing strategy and Illumina sequencing technology. The BAC library is consisted of 76,023 clones with average insert length of 123.48 Kb, about 9.9x coverage of the 850 Mb genome. Then, 8262 of 9216 clones were mapped on broomcorn millet cultivar longmi4 genome using our pipeline. Furthermore, we also extracted and assembled unmapped reads against longmi4 genome. A total of 135 deletion sequences, 64 specific sequences and some sample contamination sequences were identified.
## Conclusions
BAC clones in this library can be browsed and obtained from our website (http://eightstarsbio.com/gresource/JBrowse-1.16.5/index.html). This pipeline designed for BAC end profiling can greatly reduce the cost of acquiring BAC end sequences and shorten the period of the experiment compared with the Sanger sequencing method. These high-quality BAC clones related with genome in this study provide a useful and convinient genomic resource for genome gap filling, complex segment sequencing, FISH, functional research, and genetic engineering.


BAC clones can be browsed, searched and obtained at our website
http://eightstarsbio.com/gresource/JBrowse-1.16.5/index.html (The url is available, Sep 17, 2019.)


Our home page:http://GResource.hzau.edu.cn  http://eightstasbio.com/

This artical posted on bioarchive:  https://www.biorxiv.org/content/10.1101/2020.11.19.389643v1

Email: xuweixw27@gmail.com


#### Illumina sequencing data sets are available at Sequence Read Archive (SRA) under accession PRJNA576359
